docs/conditions.md

```diff
@@ -0,0 +1,150 @@
+# CR Conditions and Readiness Probe
+Conditions are..
+	   _the latest available observations of an object's state. They are
+	   an extension mechanism intended to be used when the details of an
+	   observation are not a priori known or would not apply to all
+	   instances of a given Kind._
+
+Kubernetes conditions [documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).
+
+The HCO’s CR is a representation of the all the underlying component operators'
+state.  In theory, if the HCO's CR exists, then all component CRs _should_
+exist, and all applications _should_ exist.  If the object doesn’t exist, then
+all component CRs _should not_ exist, and all applications _should not_ exist.
+However, the CR existence can only can tell us if the application should exist and
+doesn't help us observe the application's health. This is where the HCO and
+component operators will use conditions on their CRs to reflect the health of
+the underlying application.  Component operators store conditions that are
+watched by the HCO and the HCO will store conditions that reflect the
+[worst state](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/docs/conditions.md#hco-conditions) of all component operator conditions.
+
```

Comment:
FWIW, in a drain controller design discussion this week, I saw Doug say:

> All the advice I've seen recently says that "conditions" are no longer the preferred way to express overall status, and that explicit fields should be used instead. The reason I've seen is that having a list of conditions makes it harder to make (or maybe keep) API promises than having regular fields on the CRD.

(I don't know have any further background on this line of thinking, and Doug is away this week. So again, just FWIW)

Reply:
Interesting that you mention that. So the [api-conventions doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) is slightly older than these issues talking about Status and Conditions:

- https://github.com/kubernetes/kubernetes/issues/50798
- https://github.com/kubernetes/kubernetes/issues/7856

There do seem to be some reasonable arguments on both sides of this discussion. At the same time, this is exactly what the Cluster Version Operator is doing. Without having a convincing argument to not use Conditions or clearly established alternative, I still think Conditions are the way to move forward.

Reply:
This might be the issue where it was discussed https://github.com/kubernetes-sigs/cluster-api/pull/1081 ?

Reply:
I think what's outlined in this doc follows the "outlook" approach written [here](https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-99667941).  Since this topic is long running discussion, I'm in favor of implementing this approach and following whatever the resolution of https://github.com/kubernetes/kubernetes/issues/7856 is at a later date.